python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,515 @@
+        self.recursive = recursive
+        self.allow_non_existent = allow_non_existent
+
+    @property
+    def base_dir(self):
+        return pathlib.Path(self.selector.base_dir)
+
+    @base_dir.setter
+    def base_dir(self, base_dir):
+        self.selector.base_dir = _path_to_bytes(base_dir)
+
+    @property
+    def allow_non_existent(self):
+        return self.selector.allow_non_existent
+
+    @allow_non_existent.setter
+    def allow_non_existent(self, bint allow_non_existent):
+        self.selector.allow_non_existent = allow_non_existent
+
+    @property
+    def recursive(self):
+        return self.selector.recursive
+
+    @recursive.setter
+    def recursive(self, bint recursive):
+        self.selector.recursive = recursive
+
+
+cdef class FileSystem:
+    """Abstract file system API"""
+
+    cdef:
+        shared_ptr[CFileSystem] wrapped
+        CFileSystem* fs
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    cdef init(self, const shared_ptr[CFileSystem]& wrapped):
+        self.wrapped = wrapped
+        self.fs = wrapped.get()
+
+    def get_target_stats(self, paths_or_selector):
+        """Get statistics for the given target.
+
+        Any symlink is automatically dereferenced, recursively. A non-existing
+        or unreachable file returns an Ok status and has a FileType of value
+        NonExistent. An error status indicates a truly exceptional condition
+        (low-level I/O error, etc.).
+
+        Parameters
+        ----------
+        paths_or_selector: Union[Selector, List[Union[str, pathlib.Path]]]
+            Either a Selector object or a list of path-like objects.
+            The selector's base directory will not be part of the results, even
+            if it exists. If it doesn't exist, use `allow_non_existent`.
+
+        Returns
+        -------
+        file_stats : List[FileStats]
+        """
+        cdef:
+            vector[CFileStats] stats
+            vector[c_string] paths
+            CSelector selector
+
+        if isinstance(paths_or_selector, Selector):
+            selector = (<Selector>paths_or_selector).selector
+            check_status(self.fs.GetTargetStats(selector, &stats))
+        elif isinstance(paths_or_selector, (list, tuple)):
+            paths = [_path_to_bytes(s) for s in paths_or_selector]
+            check_status(self.fs.GetTargetStats(paths, &stats))
+        else:
+            raise TypeError('Must pass either paths or a Selector')
+
+        return [FileStats.wrap(stat) for stat in stats]
+
+    def create_dir(self, path, bint recursive=True):
+        """Create a directory and subdirectories.
+
+        This function succeeds if the directory already exists.
+
+        Parameters
+        ----------
+        path : Union[str, pathlib.Path]
+            The path of the new directory.
+        recursive: bool, default True
+            Create nested directories as well.
+        """
+        check_status(
```

Comment:
Also release the GIL here, and probably in all other FileSystem method calls below.